libs/contrib/Data/List/Lazy.idr

```diff
@@ -124,16 +124,16 @@ Monad LazyList where
 -- The result of a traversal will be a non-lazy list in general
 -- (you can't delay the "effect" of an applicative functor).
 public export
-traverse : Applicative f => (a -> f b) -> LazyList a -> f (List b)
+traverse : Monad f => (a -> f b) -> LazyList a -> f (List b)
 traverse g [] = pure []
-traverse g (x :: xs) = [| g x :: traverse g xs |]
+traverse g (x::xs) = pure $ !(g x) :: !(traverse g xs)
```

Comment:
How is this any lazier? `!` lifts the expression to the surrounding `do` block so we do run `traverse` before emitting `(::)`.

Reply:
if gives `g x >>= \y => traverse g xs >>= pure (y :: ys)`, which means that `traverse g xs` is computed only lazily demanded by `>>=`. When we use `<*>` for this, it is computed eagerly. By monadic laws, these runs give the same result, but monadic laws do not talk about laziness.

Reply:
This is the same difference as between non-lazy applicative `*>` and lazy monadic `>>`